docs/src/explanations/naming.md

```diff
@@ -0,0 +1,214 @@
+primarily relying on reflection to find names, (2) using `@chiselName` macro which had unreliable behavior, and (3)
+when all else fails using the `.suggestName("foo")` API.
+
+With Chisel 3.4 and later, a custom Scala compiler plugin was introduced which enabled reliably and automatically
+capturing the names of signals, when they are declared. In addition, this release included prolific use of a new
+prefixing API which enabled more stable naming of signals programmatically generated from function calls.
+
+This document explains how naming now works in Chisel for signal and module names. For cookbook examples on how to fix
+systemic name-stability issues, please refer to the naming [cookbook](../cookbooks/naming.md).
+
+### Compiler Plugin
+
+```scala mdoc:invisible
+import chisel3.plugin._
+import chisel3._
+import chisel3.experimental.prefix
+import chisel3.experimental.noPrefix
+import chisel3.stage.ChiselStage
+```
+
+With the release of Chisel 3.4, all users must add the following line to their build.sbt settings to get the improved
+naming:
+
+```scala
+addCompilerPlugin("edu.berkeley.cs" % "chisel3-plugin" % chiselVersion cross CrossVersion.full)
+```
+
+This plugin will run after the 'typer' phase of the Scala compiler. It looks for any user code which is of the form
+`val x = y`, where `x` is of type `chisel3.Data`, `chisel3.MemBase`, or `chisel3.experimental.BaseModule`. For each
+line which fits this criteria, it rewrites that line. In the following examples, the commented line is the what the
+line above is rewritten to.
+
+If the line is within a bundle declaration or is a module instantiation, it is rewritten to replace the right hand
+side with a call to `autoNameRecursively`, which names the signal/module.
+
+```scala mdoc
+class MyBundle extends Bundle {
+  val foo = Input(UInt(3.W))
+  // val foo = autoNameRecursively("foo", Input(UInt(3.W)))
+}
+class Example1 extends MultiIOModule {
+  val io = IO(new MyBundle())
+  // val io = autoNameRecursively("io", IO(new MyBundle()))
+}
+println(ChiselStage.emitVerilog(new Example1))
+```
+
+Otherwise, it is rewritten to also include the name as a prefix to any signals generated while executing the right-hand-
+side of the val declaration:
+
+```scala mdoc
+class Example2 extends MultiIOModule {
+  val in = IO(Input(UInt(2.W)))
+  // val in = autoNameRecursively("in", prefix("in")(IO(Input(UInt(2.W)))))
+
+  val out = IO(Output(UInt(2.W)))
+  // val out = autoNameRecursively("out", prefix("out")(IO(Output(UInt(2.W)))))
+
+  def inXin() = in * in
+
+  val add = 3.U + inXin()
+  // val add = autoNameRecursively("add", prefix("add")(3.U + inXin()))
+  // Note that the intermediate result of the first addition is prefixed with `add`
+
+  out := add + 1.U
+}
+
+println(ChiselStage.emitVerilog(new Example2))
+```
+
+Note that the naming also works if the hardware type is nested in an `Option` or an `Iterable`:
```

Comment:
```suggestion
Note that the naming also works if the hardware type is nested in an `Option` or a subtype of `Iterable`:
```
Might help people less familiar with the Scala collections library. We could also have a reference to the ScalaDoc here but not needed.